Sources/GRPC/ConnectionPool/ConnectionPool.swift

```diff
@@ -0,0 +1,660 @@
+      promise.futureResult.whenComplete { _ in
+        self.state = .shutdown
+        self.logger.trace("finished shutting down connection pool")
+      }
+
+      // Shutdown all the connections and remove them from the pool.
+      let allShutdown: [EventLoopFuture<Void>] = self.connections.values.map {
+        return $0.manager.shutdown()
+      }
+      self.connections.removeAll()
+
+      // Fail the outstanding waiters.
+      while var waiter = self.waiters.popFirst() {
+        waiter.fail(ConnectionPoolError.shutdown)
+      }
+
+      // Cascade the result of the shutdown into the promise.
+      EventLoopFuture.andAllSucceed(allShutdown, promise: promise)
+
+    case let .shuttingDown(future):
+      // We're already shutting down, cascade the result.
+      promise.completeWith(future)
+
+    case .shutdown:
+      // Already shutdown, fine.
+      promise.succeed(())
+    }
+  }
+}
+
+extension ConnectionPool: ConnectionManagerConnectivityDelegate {
+  func connectionStateDidChange(
+    _ manager: ConnectionManager,
+    from oldState: ConnectivityState,
+    to newState: ConnectivityState
+  ) {
+    switch (oldState, newState) {
+    case (.connecting, .ready):
+      // This is the only valid transition to ready.
+      self.connectionAvailable(manager.id)
+
+    case (.ready, .transientFailure),
+         (.ready, .idle),
+         (.ready, .shutdown):
+      // The connection is no longer available.
+      self.connectionUnavailable(manager.id)
+
+    default:
+      // We're only interested in becoming 'ready' and no longer being 'ready'.
+      ()
+    }
+  }
+
+  func connectionIsQuiescing(_ manager: ConnectionManager) {
+    self.eventLoop.assertInEventLoop()
+    guard let removed = self.connections.removeValue(forKey: manager.id) else {
+      return
+    }
+
+    // Drop any delegates. We're no longer interested in these events.
+    removed.manager.sync.connectivityDelegate = nil
+    removed.manager.sync.http2Delegate = nil
+
+    // Replace the connection with a new idle one.
+    self.addConnectionToPool()
+
+    // Since we're removing this connection from the pool, the pool manager can ignore any streams
+    // reserved against this connection.
+    //
+    // Note: we don't need to adjust the number of available streams as the number of connections
+    // hasn't changed.
+    self.streamLender.returnStreams(removed.reservedStreams, to: self)
+  }
+
+  /// A connection is now in the ready state, it can provide streams to waiters.
+  private func connectionAvailable(_ id: ConnectionManagerID) {
+    self.eventLoop.assertInEventLoop()
+    self.connections[id]?.available(maxConcurrentStreams: self.maxConcurrentStreams)
+
+    // A connection becoming ready is usually followed by an update to max concurrent streams.
+    // Instead of servicing waiters right now we'll do this in a jiffy, i.e. once we receive the
+    // update to max concurrent streams. This allows us to avoid succeeding too many waiters if max
```

Comment:
This seems a little brittle: what if we _don't_ receive the update to max concurrent streams in a timely fashion? Would it be better to trigger this based on the receipt of that frame?

Reply:
I also didn't like this -- feels like a massive hack. I wasn't sure about receiving the update though. IIRC the settings frame after the connection preface is allowed to be empty, what can we do if this is the case, can we assume a value of max concurrent streams and go about our business until we know better or do we have to wait until we have a value before opening streams?

Reply:
Maybe this isn't a problem in practice though 🤷‍♂️

Reply:
If the settings is empty then the default value applies, we are using the heuristic that the value won't change.

Reply:
Okay, that makes sense, I'll incorporate this suggestion. I think emitting an update of the default value if we don't receive one in the first settings frame should work fine.